design/expected-hardware-configuration-validation.md

```diff
@@ -0,0 +1,262 @@
+
+## Proposal
+
+We are creating proposal based on the discussion with metal3 community on issue #351:
+https://github.com/metal3-io/baremetal-operator/issues/351
+
+We compared introspection data of Ironic with metal3 schema and we found that default  hardware
+configuration needs to be added by introducing a new CRD `HardwareClassificationController`.
+
+We will write a CRD that checks inspected baremetal hosts against hardware configuration and add label as "BMO-OK" for matched profile into baremetal host baremetalHost_crd.yaml.
+  
+### Implementation Details/Notes/Constraints
+
+Link for Existing Metal3 Specs
+Please refer metal3 spec for bare-metal:
+https://github.com/metal3-io/baremetal-operator/blob/master/deploy/crds/metal3.io_baremetalhosts_crd.yaml
+
+* Write a below schema for new CRD under folder deploy/crds for Kind HardwareClassificationController.
+
+    ```yaml
+    - properties:
+        expectedHardwareConfiguration:
+            default:
+            properties:
+            customFilter:
+                    type: String
+            namespace:
+                    type: String
+            profileName:
+                    type: String
+            cpu:
+                properties:
+                    count:
+                        type: Integer
+            type: Object
+            disk:
+                properties:
+                    rootDisk:
+                        sizeBytesGB:
+                            type: Integer
+                    type: Object
+                    dataDisk:
+                        sizeBytesGB:
+                            type: Integer
+                    numberOfDisks:
+                        type: Integer
+                    type: array
+            type: Object
+            nics:
+                properties:
+                    numberOfNICS:
+                        type: Integer
+            type: Object
+            ram:
+                properties:
+                    sizeBytesGB:
+                        type: Integer
+            type: Object
+            systemVendor:
+                properties:
+                    productName:
+                        type: String
+                    manufacturer:
+                        type: String
+            type: Object
+            firmware:
+                properties:
+                    version:
+                        RAID:
+                            type: String
+                        BasebandManagement:
+                            type: String
+                        BIOS:
+                            type: String
+                        IDRAC:
+                            type: String
+                type: Object
+            required:
+            - customFilter
+            - namespace
+            - profileName
+        rules:
```

Comment:
We don't need both the `expectedHardwareConfiguration` section and the `rules` section. The rules were meant to allow us to describe everything in the hardware configuration section more generally. We should pick one approach.

I proposed the general rules form, but I'm not set on that as a requirement if we can make the other form reasonably expressive. Let's see if we can work on that by writing out some example profiles.

Reply:
Okay. We can discuss this thing in our call.